src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,180 @@
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // To keep track of where we are and how things are going.
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Initialize the update policy.
+  if (resetPolicy)
+    updatePolicy.Initialize(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+               std::numeric_limits<size_t>::max() : maxIterations;
+
+  // initialize the gradients.
+  size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+  arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+  arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+  for (size_t f = 0, b=0; f < numFunctions;
+    /* incrementing done manually */)
+  {
+    // Find the effective batch size (the last batch may be smaller).
+    effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+    function.Gradient(iterate, f, tableOfGradients.slice(b),
+                      effectiveBatchSize);
+
+    f += effectiveBatchSize;
+    avgGradient += tableOfGradients.slice(b);
+    b++;
+  }
+  avgGradient /= (double) numBatches; // average gradient
+
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "SAGA: converged to " << overallObjective
+                << "; terminating  with failure.  Try a smaller step size?"
+                << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SAGA: minimized within tolerance " << tolerance
+                << "; terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    lastObjective = overallObjective;
+
+    // Store current parameter for the calculation of the variance reduced
+    // gradient.
+    iterate0 = iterate;
+
+    for (size_t f = 0, b=0, currentFunction = 0; f < numFunctions;
+      /* incrementing done manually */)
+    {
+      b = math::RandInt(0, numBatches); // Random batch selected
+      // Find the effective batch size (the last batch may be smaller).
+      currentFunction = b*batchSize;
```

Comment:
Please use a space between operations, there are more examples here, so please check the code for the pattern.